File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] inputNo = br.readLine().trim().split(" ");
		
		int M = Integer.parseInt(inputNo[0]);
		int N = Integer.parseInt(inputNo[1]);
		long[][] input = new long[M][M];
		long[][] input_sum = new long[M][M];
		for(int i=0; i<M; i++) {
			inputNo = br.readLine().trim().split(" ");
			for(int j=0; j<M; j++) {
				input[i][j]=Long.parseLong(inputNo[j]);
				if(i!=0) {
					if(j==0) {
						input_sum[i][j]= (long) input[i][j] + input_sum[i-1][j];
					}else {						
						input_sum[i][j]=(long) input_sum[i][j-1]+input[i][j]+input_sum[i-1][j]-input_sum[i-1][j-1];
					}
				}else {
					if(j==0) {
						input_sum[i][j]= (long) input[i][j];
					}else {						
					input_sum[i][j]=(long) input_sum[i][j-1]+input[i][j];
					}
				}
						
			}
		}
		for(int i=0; i<N; i++) {
			long sum = 0;
			inputNo = br.readLine().trim().split(" ");
			int x_1 = Integer.parseInt(inputNo[0]);
			int y_1 = Integer.parseInt(inputNo[1]);
			int x_2 = Integer.parseInt(inputNo[2]);
			int y_2 = Integer.parseInt(inputNo[3]);
			if(x_1>1 && y_1>1) {
				sum = (long) input_sum[x_2-1][y_2-1]-input_sum[x_1-2][y_2-1]-input_sum[x_2-1][y_1-2]+input_sum[x_1-2][y_1-2];
				
			}else if(x_1==1 && y_1>1){
				sum = (long) input_sum[x_2-1][y_2-1]-input_sum[x_2-1][y_1-2];
				
			}else if(x_1>1 && y_1==1){
				sum = (long) input_sum[x_2-1][y_2-1]-input_sum[x_1-2][y_2-1];
			}else if(x_1==1 && y_1==1){
				sum = (long) input_sum[x_2-1][y_2-1];
			}	
					
				
			System.out.println(sum);
		}

	}

}

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.py
M, N = map(int, input().split())
matrix = [[0]*(M+1)]+[[0]+list(map(int, input().split())) for _ in range(M)]
# for i in range(M):
#     print(matrix[i])
# print()
sum_matrix = [[0]*(M+1) for _ in range(M+1)]
result_matrix = [list(map(int,input().split())) for _ in range(N)]
for i in range(1,M+1):
    sum_matrix[i][1]=sum_matrix[i-1][1]+matrix[i][1]
    for j in range(2,M+1):
        sum_matrix[i][j] = sum_matrix[i][j-1]+sum_matrix[i-1][j]+matrix[i][j]-sum_matrix[i-1][j-1]
# for i in range(M+1):
#     print(sum_matrix[i])
for i in range(N):
    s,e,S,E = result_matrix[i][0], result_matrix[i][1], result_matrix[i][2], result_matrix[i][3]
    print(sum_matrix[S][E]-sum_matrix[S][e-1]-sum_matrix[s-1][E]+sum_matrix[s-1][e-1])
